<?xml version="1.0"?>
<launch>

	<!--user configure parameters for ros start-->
	<arg name="lvx_file_path" default="livox_test.lvx"/>
	<arg name="bd_list" default="100000000000000"/>
	<arg name="xfer_format" default="1"/>
	<arg name="multi_topic" default="0"/>
	<arg name="data_src" default="0"/>
	<arg name="publish_freq" default="10.0"/>
	<arg name="output_type" default="0"/>
	<arg name="rviz_enable" default="false"/>
	<arg name="rosbag_enable" default="false"/>
	<arg name="cmdline_arg" default="$(arg bd_list)"/>
	<arg name="msg_frame_id" default="livox_frame"/>
	<arg name="lidar_bag" default="false"/>
	<arg name="imu_bag" default="false"/>
	<arg name="sample_at_startup" default="false"/>
	<arg name="config_file" default="$(find livox_ros_driver2)/config/MID360_config.json"/>
	<arg name="enable_box_filtering" default="true"/>
	<arg name="enable_tag_filtering" default="true"/>
	<arg name="box_frame" default="aristos_base_footprint"/>
	<arg name="box_min_x" default="-1.0"/>
	<arg name="box_min_y" default="-1.0"/>
	<arg name="box_min_z" default="-1.0"/>
	<arg name="box_max_x" default="1.0"/>
	<arg name="box_max_y" default="1.0"/>
	<arg name="box_max_z" default="1.0"/>
	<arg name="other_high" default="true"/>
	<arg name="other_moderate" default="true"/>
	<arg name="other_low" default="true"/>
	<arg name="other_rsrv" default="true"/>
	<arg name="atm_high" default="true"/>
	<arg name="atm_moderate" default="true"/>
	<arg name="atm_low" default="true"/>
	<arg name="atm_rsrv" default="true"/>
	<arg name="drag_high" default="true"/>
	<arg name="drag_moderate" default="true"/>
	<arg name="drag_low" default="true"/>
	<arg name="drag_rsrv" default="true"/>
	<arg name="rsrv_zero" default="true"/>
	<arg name="rsrv_one" default="true"/>
	<arg name="rsrv_two" default="true"/>
	<arg name="rsrv_three" default="true"/>
	<arg name="pcl_topic_timeout" default="10.0"/>
	<arg name="user_few_points_number_threshold" default="4000"/>
	<arg name="user_few_points_time_threshold" default="10.0"/>
	
	<!--user configure parameters for ros end--> 

	<!-- Nodelet manager -->
    <node pkg="nodelet" type="nodelet" name="lidar_nodelet_manager" args="manager" output="screen"/>

    <!-- New nodelet -->
    <node pkg="nodelet" type="nodelet" name="livox_lidar_publisher2" args="load livox_ros_driver2/DriverNode lidar_nodelet_manager" output="screen">
        <remap from="livox_lidar_publisher2/livox/enable_sampling" to="livox/enable_sampling"/>
		<remap from="livox_lidar_publisher2/livox/lidar" to="livox/lidar"/>
		<remap from="livox_lidar_publisher2/livox/imu" to="livox/imu"/>
		<remap from="livox_lidar_publisher2/livox/info" to="livox/info"/>
		<param name="xfer_format" value="$(arg xfer_format)"/>
		<param name="multi_topic" value="$(arg multi_topic)"/>
		<param name="data_src" value="$(arg data_src)"/>
		<param name="publish_freq" type="double" value="$(arg publish_freq)"/>
		<param name="output_data_type" value="$(arg output_type)"/>
		<param name="cmdline_str" type="string" value="$(arg bd_list)"/>
		<param name="cmdline_file_path" type="string" value="$(arg lvx_file_path)"/>
		<param name="user_config_path" type="string" value="$(arg config_file)"/>
		<param name="frame_id" type="string" value="$(arg msg_frame_id)"/>
		<param name="enable_lidar_bag" type="bool" value="$(arg lidar_bag)"/>
		<param name="enable_imu_bag" type="bool" value="$(arg imu_bag)"/>
		<param name="sample_at_startup" type="bool" value="$(arg sample_at_startup)"/>
    </node>

	<node pkg="nodelet" type="nodelet" name="custom_livox_filter" args="load livox_ros_driver2/CustomFilterNodelet lidar_nodelet_manager" output="screen">
		<remap from="input_point_cloud" to="livox/lidar"/>
		<remap from="filtered_point_cloud" to="livox/lidar_filtered"/>
		<param name="enable_box_filtering" type="bool" value="$(arg enable_box_filtering)"/>
		<param name="enable_tag_filtering" type="bool" value="$(arg enable_tag_filtering)"/>
		<param name="box_frame" type="string" value="$(arg box_frame)"/>
		<param name="box_min_x" type="double" value="$(arg box_min_x)"/>
		<param name="box_min_y" type="double" value="$(arg box_min_y)"/>
		<param name="box_min_z" type="double" value="$(arg box_min_z)"/>
		<param name="box_max_x" type="double" value="$(arg box_max_x)"/>
		<param name="box_max_y" type="double" value="$(arg box_max_y)"/>
		<param name="box_max_z" type="double" value="$(arg box_max_z)"/>
		<param name="other_high" type="bool" value="$(arg other_high)"/>
		<param name="other_moderate" type="bool" value="$(arg other_moderate)"/>
		<param name="other_low" type="bool" value="$(arg other_low)"/>
		<param name="other_rsrv" type="bool" value="$(arg other_rsrv)"/>
		<param name="atm_high" type="bool" value="$(arg atm_high)"/>
		<param name="atm_moderate" type="bool" value="$(arg atm_moderate)"/>
		<param name="atm_low" type="bool" value="$(arg atm_low)"/>
		<param name="atm_rsrv" type="bool" value="$(arg atm_rsrv)"/>
		<param name="drag_high" type="bool" value="$(arg drag_high)"/>
		<param name="drag_moderate" type="bool" value="$(arg drag_moderate)"/>
		<param name="drag_low" type="bool" value="$(arg drag_low)"/>
		<param name="drag_rsrv" type="bool" value="$(arg drag_rsrv)"/>
		<param name="rsrv_zero" type="bool" value="$(arg rsrv_zero)"/>
		<param name="rsrv_one" type="bool" value="$(arg rsrv_one)"/>
		<param name="rsrv_two" type="bool" value="$(arg rsrv_two)"/>
		<param name="rsrv_three" type="bool" value="$(arg rsrv_three)"/>
	</node>

	<node pkg="nodelet" type="nodelet" name="pcl_analyzer" args="load livox_ros_driver2/PclAnalyzerNodelet lidar_nodelet_manager" output="screen">
		<remap from="input_point_cloud" to="livox/lidar_filtered"/>
		<remap from="active" to="livox/active"/>
		<remap from="enough" to="livox/enough_points"/>
		<param name="msg_type" type="int" value="$(arg xfer_format)"/>
		<param name="pcl_topic_timeout" type="double" value="$(arg pcl_topic_timeout)"/>
		<param name="user_few_points_number_threshold" type="int" value="$(arg user_few_points_number_threshold)"/>
		<param name="user_few_points_time_threshold" type="double" value="$(arg user_few_points_time_threshold)"/>
	</node>

	<group if="$(arg rviz_enable)">
		<node name="livox_rviz" pkg="rviz" type="rviz" respawn="true"
				args="-d $(find livox_ros_driver2)/config/display_point_cloud_ROS1.rviz"/>
    </group>


</launch>
